### PR TITLE
Inventory Update

### PR DESCRIPTION
**Pull request by Molkars**
Add models for all of the inventory, project, and time-management features
Add controllers for all inventory models
Added optional count and offset queries to get array queries

### CODE REVIEW

File: bin/generate_client.dart [hunk 1]
```diff
@@ -0,0 +1,130 @@
+import 'dart:io';
```
Comment (brandan-schmitz) [inline bin/generate_client.dart:1]: Lets add this file to the gitignore file. We do not need this in the repository.

Comment (Molkars) [reply bin/generate_client.dart:1]: I didn't mean to push that to begin with

**Review by brandan-schmitz — CHANGES_REQUESTED** [review]
There are a few changes that should get made before committing them. Use this review as a checklist to complete them. Mark each review comment as resolved as once they have been updated.

File: lib/controller/user_controller.dart [hunk 2]
```diff
@@ -78,20 +78,13 @@ class UserController extends ResourceController {
...
-    });
 
     if (operation.pathVariables.isEmpty && operation.method == "GET") {
-      _response['200'] = APIResponse.schema("Users Fetched Successfully", APISchemaObject.array(ofSchema: _userSchema));
+      _response['200'] = APIResponse.schema("Users Fetched Successfully", APISchemaObject.array(ofSchema: context.schema.getObjectWithType(User)));
     } else if (operation.pathVariables.isNotEmpty && operation.method == "GET") {
-      _response['200'] = APIResponse.schema("User Fetched Successfully", _userSchema);
+      _response['200'] = APIResponse.schema("User Fetched Successfully", context.schema.getObjectWithType(User));
     } else if (operation.pathVariables.isNotEmpty && operation.method == "PATCH") {
-      _response['200'] = APIResponse.schema("User Updated Successfully", _userSchema);
```
Comment (brandan-schmitz) [inline lib/controller/user_controller.dart:94]: We should not be returning the user tokens and the user's passwordResetTokens so those need to get removed from the Context of the user. This should get reverted to what it was before.

**Review by brandan-schmitz — CHANGES_REQUESTED** [review]

File: lib/controller/password_reset_controller.dart [hunk 3]
```diff
@@ -142,15 +142,22 @@ class PasswordResetController extends ResourceController {
...
+            '200': APIResponse.schema("Request successful", null),
+            '404': APIResponse.schema("That email is not associated with a user account", APISchemaObject.empty()),
+          };
+        } else {
+          return {
+            '200': APIResponse.schema("Password Reset successful", null),
+            '403': APIResponse.schema("Token has expired", APISchemaObject.empty()),
+            '404': APIResponse.schema("Invalid token", APISchemaObject.empty()),
+          };
+        }
```
[the remark this answers is not in the document]
Comment (Molkars) [reply lib/controller/password_reset_controller.dart:158]: The case 'email' will never occur, email is query parameter not a path variable and will never appear in operation.pathVariables.

File: lib/controller/register_controller.dart [hunk 4]
```diff
@@ -5,11 +5,16 @@ class RegisterController extends ResourceController {
...
+  /// Returns a 500 BAD REQUEST if any of the username, password, firstName, lastName,
+  /// email, or id is null
+  /// The password is hashed
   @Operation.post()
   Future<Response> createUser(@Bind.body(ignore: ["id"]) User user) async {
-    // Check for required parameters before we spend time hashing
+    // Check for @required parameters before we spend time hashing
     if (user.username == null || user.password == null || user.firstName == null || user.lastName == null || user.email == null) {
-      return Response.badRequest(body: {"error": "username, password, firstName, lastName, and email fields are required."});
+      return Response.badRequest(body: {"error": "username, password, firstName, lastName, and email fields are @required."});
```
[the remark this answers is not in the document]
Comment (Molkars) [reply lib/controller/register_controller.dart:17]: I agree, that was an unforseen consequence of a mass-replace I did

**Pull request closed without merging**